### PR TITLE
Smallrobotic

### PR DESCRIPTION
**Pull request by MichaelCourtney**
https://www.freeorion.org/forum/viewtopic.php?f=15&t=12234https://www.freeorion.org/forum/viewtopic.php?f=15&t=12234

### CODE REVIEW

**Comment by MichaelCourtney** [comment]
Lacks a proper graphic (its currently on the generic). Fluff and position in the tech tree, etc. could do with a looking over.

**Comment by MichaelCourtney** [comment]
The fail is just checksums that need to be changed? Rather then something more serious?

**Comment by geoffthemedio** [comment]
> The fail is just checksums that need to be changed? Rather then something more serious?

I think so.

File: test/system/TestChecksum.cpp [hunk 1]
```diff
@@ -41,17 +41,17 @@ void TestCheckSumFromEnv(const char* env, unsigned int def, unsigned int calcula
...
-    TestCheckSumFromEnv("FO_CHECKSUM_BUILDING", 7870370, checksums["BuildingTypeManager"]);
+    TestCheckSumFromEnv("FO_CHECKSUM_BUILDING", 9738579, checksums["BuildingTypeManager"]);
     TestCheckSumFromEnv("FO_CHECKSUM_ENCYCLOPEDIA", 1090024, checksums["Encyclopedia"]);
-    TestCheckSumFromEnv("FO_CHECKSUM_FIELD", 3780088, checksums["FieldTypeManager"]);
-    TestCheckSumFromEnv("FO_CHECKSUM_POLICY", 7303441, checksums["PolicyManager"]);
-    TestCheckSumFromEnv("FO_CHECKSUM_SHIP_HULL", 5932838, checksums["ShipHullManager"]);
-    TestCheckSumFromEnv("FO_CHECKSUM_SHIP_PART", 1765649, checksums["ShipPartManager"]);
+    TestCheckSumFromEnv("FO_CHECKSUM_FIELD", 5633722, checksums["FieldTypeManager"]);
+    TestCheckSumFromEnv("FO_CHECKSUM_POLICY", 7039723, checksums["PolicyManager"]);
+    TestCheckSumFromEnv("FO_CHECKSUM_SHIP_HULL", 8777190, checksums["ShipHullManager"]);
```
Comment (agrrr3) [inline test/system/TestChecksum.cpp:48]: ship_hull checksum change is expected here; seems i cant update the pull request with the current values from the fail; from the log
873430

File: test/system/TestChecksum.cpp [hunk 2]
```diff
@@ -41,17 +41,17 @@ void TestCheckSumFromEnv(const char* env, unsigned int def, unsigned int calcula
...
+    TestCheckSumFromEnv("FO_CHECKSUM_SHIP_PART", 1778809, checksums["ShipPartManager"]);
     TestCheckSumFromEnv("FO_CHECKSUM_SHIP_DESIGN", 872567, checksums["PredefinedShipDesignManager"]);
-    TestCheckSumFromEnv("FO_CHECKSUM_SPECIES", 181788, checksums["SpeciesManager"]);
-    TestCheckSumFromEnv("FO_CHECKSUM_SPECIALS", 9623845, checksums["SpecialsManager"]);
-    TestCheckSumFromEnv("FO_CHECKSUM_TECH", 5378437, checksums["TechManager"]);
-    TestCheckSumFromEnv("FO_CHECKSUM_NAMED_VALUEREF", 8513318, checksums["NamedValueRefManager"]);
+    TestCheckSumFromEnv("FO_CHECKSUM_SPECIES", 3127776, checksums["SpeciesManager"]);
+    TestCheckSumFromEnv("FO_CHECKSUM_SPECIALS", 7644502, checksums["SpecialsManager"]);
+    TestCheckSumFromEnv("FO_CHECKSUM_TECH", 5478559, checksums["TechManager"]);
```
Comment (agrrr3) [inline test/system/TestChecksum.cpp]: tech checksum change expected; update from fail log:
5486989

File: test/system/TestChecksum.cpp (hunk 1 above)
Comment (MichaelCourtney) [reply test/system/TestChecksum.cpp:48]: Thanks. I've updated. Looking in the error log I'm pretty sure this one was ship_design rather then ship_hull

**Comment by MichaelCourtney** [comment]
Hopefully that's got it this time. The tech size in the python parser was different, as I'd added a tech. If it fails this time, I'm out of ideas

**Comment by o01eg** [comment]
It would be nice if that tech will be written in Python FOCS already. It doesn't have effects so could be simple.

**Comment by MichaelCourtney** [comment]
> It would be nice if that tech will be written in Python FOCS already. It doesn't have effects so could be simple.

I'll take a look at 1 you've already done tomorrow and rewrite it.

**Comment by MichaelCourtney** [comment]
@o01eg You might need to look over it and see if I've done this right

**Comment by o01eg** [comment]
Oh, I'm not yet implemented `ShipHull` token for unlock.

**Comment by o01eg** [comment]
I've added support for it in #3983.

**Comment by o01eg** [comment]
@MichaelCourtney I've merged it, token for unlock is `UnlockShipHull`.

**Comment by o01eg** [comment]
It's fine now.

**Comment by MichaelCourtney** [comment]
Well it was till I tried squishing the commits...

**Comment by agrrr3** [comment]
> Well it was till I tried squishing the commits...

? well, you can always `git reset --hard 263e1be` and start squishing again

**Comment by MichaelCourtney** [comment]
Ok now I'm confused. It was working, I squished the commits, got a fail message, went out, came home to a green tick and "all checks have passed".

Seems ready?

**Comment by agrrr3** [comment]
i am a bit confused about the checksums. You added a tech and a hull, and moved NamedReal. 
So TechManager, ShipHullManager, (very maybe) NamedValueRefManager, (very maybe) ShipDesignManager and (maybe) Encyclopedia checksums should have changed.

But nothing else (e.g. BuildingTypeManager) should have changed checksums.

From test/system/TestChecksum.cpp in master:
```    TestCheckSumFromEnv("FO_CHECKSUM_BUILDING", 7870370, checksums["BuildingTypeManager"]);
    TestCheckSumFromEnv("FO_CHECKSUM_ENCYCLOPEDIA", 1090024, checksums["Encyclopedia"]);
    TestCheckSumFromEnv("FO_CHECKSUM_FIELD", 3780088, checksums["FieldTypeManager"]);
    TestCheckSumFromEnv("FO_CHECKSUM_POLICY", 7303441, checksums["PolicyManager"]);
    TestCheckSumFromEnv("FO_CHECKSUM_SHIP_HULL", 5932838, checksums["ShipHullManager"]);
    TestCheckSumFromEnv("FO_CHECKSUM_SHIP_PART", 1765649, checksums["ShipPartManager"]);
    TestCheckSumFromEnv("FO_CHECKSUM_SHIP_DESIGN", 872567, checksums["PredefinedShipDesignManager"]);
    TestCheckSumFromEnv("FO_CHECKSUM_SPECIES", 181788, checksums["SpeciesManager"]);
    TestCheckSumFromEnv("FO_CHECKSUM_SPECIALS", 9623845, checksums["SpecialsManager"]);
    TestCheckSumFromEnv("FO_CHECKSUM_TECH", 5378437, checksums["TechManager"]);
    TestCheckSumFromEnv("FO_CHECKSUM_NAMED_VALUEREF", 8513318, checksums["NamedValueRefManager"]);
```

**Comment by o01eg** [comment]
Test doesn't enforce values here by default.

**Comment by MichaelCourtney** [comment]
> i am a bit confused about the checksums. You added a tech and a hull, and moved NamedReal. So TechManager, ShipHullManager, (very maybe) NamedValueRefManager, (very maybe) ShipDesignManager and (maybe) Encyclopedia checksums should have changed.

My suspicion is that some commits don't trigger all the tests and a bunch of those are commits already in master. For example I'm almost sure Encyclopedia checksums was the added pedia page over here: https://github.com/freeorion/freeorion/pull/3969

Notice how that pull request says "6 checks passed"? This pull request says "19 checks passed"

**Comment by agrrr3** [comment]
So the theory is that the checksums in master are currently wrong?

**Comment by MichaelCourtney** [comment]
> So the theory is that the checksums in master are currently wrong?

Yes. Though I'm only guessing. It certainly makes more sense for the shipdesign checksum to be the added tentacle on white krakens and the field checksum to be Oberlus's addition of named values in fields.

**Comment by geoffthemedio** [comment]
If you look in the master test build logs, eg. https://github.com/freeorion/freeorion/runs/7506756131?check_suite_focus=true you can see the checksum test results, eg. https://github.com/freeorion/freeorion/runs/7506756131?check_suite_focus=true#step:13:3931 
````
5: /home/runner/work/freeorion/freeorion/test/system/TestChecksum.cpp:22: error: warning: in "TestChecksum/compare_checksum": FO_CHECKSUM_BUILDING expected 7870370 was 9738579
5: /home/runner/work/freeorion/freeorion/test/system/TestChecksum.cpp:22: error: warning: in "TestChecksum/compare_checksum": FO_CHECKSUM_ENCYCLOPEDIA expected 1090024 was 1099081
5: /home/runner/work/freeorion/freeorion/test/system/TestChecksum.cpp:22: error: warning: in "TestChecksum/compare_checksum": FO_CHECKSUM_FIELD expected 3780088 was 5633722
5: /home/runner/work/freeorion/freeorion/test/system/TestChecksum.cpp:22: error: warning: in "TestChecksum/compare_checksum": FO_CHECKSUM_POLICY expected 7303441 was 7039723
5: /home/runner/work/freeorion/freeorion/test/system/TestChecksum.cpp:22: error: warning: in "TestChecksum/compare_checksum": FO_CHECKSUM_SHIP_HULL expected 5932838 was 5949918
5: /home/runner/work/freeorion/freeorion/test/system/TestChecksum.cpp:22: error: warning: in "TestChecksum/compare_checksum": FO_CHECKSUM_SHIP_PART expected 1765649 was 1778809
5: /home/runner/work/freeorion/freeorion/test/system/TestChecksum.cpp:22: error: warning: in "TestChecksum/compare_checksum": FO_CHECKSUM_SHIP_DESIGN expected 872567 was 873430
5: /home/runner/work/freeorion/freeorion/test/system/TestChecksum.cpp:22: error: warning: in "TestChecksum/compare_checksum": FO_CHECKSUM_SPECIES expected 181788 was 9777257
5: /home/runner/work/freeorion/freeorion/test/system/TestChecksum.cpp:22: error: warning: in "TestChecksum/compare_checksum": FO_CHECKSUM_SPECIALS expected 9623845 was 8161983
5: /home/runner/work/freeorion/freeorion/test/system/TestChecksum.cpp:22: error: warning: in "TestChecksum/compare_checksum": FO_CHECKSUM_TECH expected 5378437 was 9495557
5: /home/runner/work/freeorion/freeorion/test/system/TestChecksum.cpp:22: error: warning: in "TestChecksum/compare_checksum": FO_CHECKSUM_NAMED_VALUEREF expected 8513318 was 2621528
````

**Comment by MichaelCourtney** [comment]
> If you look in the master test build logs, eg. https://github.com/freeorion/freeorion/runs/7506756131?check_suite_focus=true you can see the checksum test results, eg. https://github.com/freeorion/freeorion/runs/7506756131?check_suite_focus=true#step:13:3931

Thanks Geoff. Anything I should do with this?

**Comment by MichaelCourtney** [comment]
@TheSilentOne1 - Hi not sure if you have much free time at the moment. Are you able to provide a graphic? It's just using the generic ship for the moment

**Comment by geoffthemedio** [comment]
> Thanks Geoff. Anything I should do with this?

You can update the checksums in your pull request so that it passes them. Not necessary, though.

**Pull request merged**